src/api-service/__app__/agent_events/__init__.py

```diff
@@ -47,11 +47,22 @@ def get_node_checked(machine_id: UUID) -> Node:
 def on_state_update(
     machine_id: UUID,
     state_update: NodeStateUpdate,
-) -> func.HttpResponse:
+) -> None:
     state = state_update.state
     node = get_node_checked(machine_id)
 
-    if state == NodeState.init or node.state not in NodeState.ready_for_reset():
+    if state == NodeState.free and (node.reimage_requested or node.delete_requested):
+        logging.info("stopping free node with reset flags: %s", node.machine_id)
+        node.stop()
+        return
+
+    if state == NodeState.init:
+        if node.delete_requested:
+            node.stop()
+            return
+        node.reimage_requested = False
+        node.save()
```

Comment:
I think this logic should be in the node class.

Reply:
probably, but I will do that in another PR